### PR TITLE
Adding labels/taints support, allow for adding more than one node pool at cluster creation and use latest Nvidia driver

### PR DESCRIPTION
**Pull request by vrutz**
Re-introducing labels and taints support for EKS node groups, updating Nvidia driver to latest and allowing for definition of multiple node groups at cluster creation (while keeping current initial node group definition in config)

When supporting taints on node groups, we have to take into account GPU enabled groups. The Nvidia driver daemonset contains a toleration for the `nvidia.com/gpu` taint to exist on the nodes it can be applied to, but it needs to be patched for other custom taints.

Regarding taints and the autoscaler, we have to patch the autoscaler configuration with the taints of the autoscaled node pools as tolerations AND we have to add the `propagateASGTags` to the node pool configuration.
This flag will add the labels and taints of the node group to the AutoScaling Group (ASG) as tags. These tags are used to label and taint new nodes on scaling up (new nodes created).

[sc-174884]
[sc-165108]
[sc-167294]
[sc-177800]

### CODE REVIEW

**Review by amandineslx — CHANGES_REQUESTED** [review]
OK for me but version bump missing in `plugin.json`

### Tested

* Deploy DSS 12.5.2 via FM in elastic fleet
* Update plugin with archive built from PR (commit `127439e98aa2cb8310c3f46a718c8f0f43a5b9b2`)
* Create cluster with: 1 node pool simple, 1 node pool with GPU and label, 1 node pool with autoscaler and taints
    * ✅ autoscaler is deployed
    * ✅ nvidia daemon set is present for each node pool
    * ✅ nvidia daemon set version is correct
    * ✅ label is populated correctly
    * ✅ taint is populated correctly
    * ✅ all 3 node pools are created in the cluster
* ✅ release notes and version bump in changelog
* ❌ version bump in `plugin.json`?

**Comment by vrutz** [comment]
> OK for me but version bump missing in `plugin.json`
> 
> ### Tested
> * Deploy DSS 12.5.2 via FM in elastic fleet
> * Update plugin with archive built from PR (commit `127439e98aa2cb8310c3f46a718c8f0f43a5b9b2`)
> * Create cluster with: 1 node pool simple, 1 node pool with GPU and label, 1 node pool with autoscaler and taints
>   
>   * ✅ autoscaler is deployed
>   * ✅ nvidia daemon set is present for each node pool
>   * ✅ nvidia daemon set version is correct
>   * ✅ label is populated correctly
>   * ✅ taint is populated correctly
>   * ✅ all 3 node pools are created in the cluster
> * ✅ release notes and version bump in changelog
> * ❌ version bump in `plugin.json`?


We are not going to release the plugin right away and probably push it back to 13.0 and bundle it with more changes. Not sure if we should bump the version then

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,11 +1,11 @@
 # Changelog
 
-## Next version
+## Version 1.4.0 - Feature release
```
Comment (amandineslx) [inline CHANGELOG.md:3]:
```suggestion
## Next version
```

Comment (amandineslx) [reply CHANGELOG.md:3]: then, we usually label the next version as "next" and don't specify the version number until we have the complete content. At release time, we decide whether we bump major, minor or patch

**Review by amandineslx — APPROVED** [review]
OK for me

**Comment by thtrunck** [comment]
We have the question of supporting also annotation, so let's clear that up with FEs and add it if needed

**Comment by vrutz** [comment]
> We have the question of supporting also annotation, so let's clear that up with FEs and add it if needed

I think annotations could be in their own PR. We are fixing the taints issue here because it'd be a bug, but annotations deserve a new Story card and probably a new PR since this one is already doing 4 different SCs

**Review by thtrunck — CHANGES_REQUESTED** [review]
As discussed we wawnt to solve the tains+gpu issue before merging that https://app.shortcut.com/dataiku/story/177800/aks-allow-nvidia-daemon-to-be-installed-on-gpu-enabled-machines-with-taints so requesting change to not mess with bots

File: python-lib/dku_utils/taints.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    def from_json(tolerations_json):
+        return [Toleration(tol) for tol in json.loads(tolerations_json)]
+
+    @staticmethod
+    def from_dict(raw_dicts):
+        return [Toleration(raw) for raw in raw_dicts or []]
+
+    @staticmethod
+    def to_dict(tolerations):
+        return [toleration.to_dict() for toleration in tolerations]
```
Comment (amandineslx) [inline python-lib/dku_utils/taints.py:53]: there is an issue with this method
it is defined to be called with an argument and not called with this argument
I think the one you want to use is a native Python

**Review by amandineslx — CHANGES_REQUESTED** [review]
got an error starting a cluster with taints due to the `to_dict` method definition

```
[2024-04-18 09:04:53,919] [34852/MainThread] [INFO] [root] Checking if NVIDIA GPU drivers are installed with : ["kubectl", "get", "pods", "--namespace", "kube-system", "-l", "name=nvidia-device-plugin-ds", "--ignore-not-found"]
Traceback (most recent call last):
  File "/opt/dataiku-dss-12.6.0/python/dataiku/cluster/server.py", line 50, in serve
    result = cluster.start()
  File "/tmp/tmp_folder_RCmKAdWZ/dku_code.py", line 272, in start
  File "/data/dataiku/dss_data/plugins/installed/eks-clusters/python-lib/dku_kube/gpu_driver.py", line 39, in add_gpu_driver_if_needed
    nvidia_config['spec']['template']['spec']['tolerations'] = Toleration.to_dict(tolerations)
  File "/data/dataiku/dss_data/plugins/installed/eks-clusters/python-lib/dku_utils/taints.py", line 53, in to_dict
    return [toleration.to_dict() for toleration in tolerations]
  File "/data/dataiku/dss_data/plugins/installed/eks-clusters/python-lib/dku_utils/taints.py", line 53, in <listcomp>
    return [toleration.to_dict() for toleration in tolerations]
TypeError: to_dict() missing 1 required positional argument: 'tolerations'
```

File: python-lib/dku_kube/autoscaler.py [hunk 3]
```diff
@@ -19,29 +20,59 @@ def has_autoscaler(kube_config_path):
...
+        logging.debug('Autoscaler deployment config: %s' % yaml.safe_dump(autoscaler_config, default_flow_style=False))
+
+    autoscaler_full_config.append(autoscaler_config)
+    logging.debug('Autoscaler complete config: %s' % yaml.safe_dump_all(autoscaler_full_config, default_flow_style=False))
+
+    with open(autoscaler_file_path, "w") as f:
+        yaml.safe_dump_all(autoscaler_full_config, f, explicit_start=True)
+
+    cmd = ['kubectl', 'apply', '-f', os.path.abspath(autoscaler_file_path)]
+    logging.info("Create autoscaler with : %s" % json.dumps(cmd))
```
[the remark this answers is not in the document]
Comment (vrutz) [reply python-lib/dku_kube/autoscaler.py:69]: Yes. I just wonder if we need to update the autoscaler pod at all. If we create a new node pool with taints, they will be propagated to the autoscaler group with the `propagateASGTags` flag. But the autoscaler will still be able to run correctly I feel.

[the remark this answers is not in the document]
Comment (vrutz) [reply python-lib/dku_kube/autoscaler.py]: I'm removing the cluster update when the autoscaler is already present for the reasons above

File: python-lib/dku_utils/taints.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        return hash((self.get('key', ''), self.get('value', ''), self.get('effect', '')))
+
+class Toleration(Taint):
+    def __init__(self, taint):
+        super().__init__(taint)
```
[the remark this answers is not in the document]
Comment (vrutz) [reply python-lib/dku_utils/taints.py:26]: We create a toleration from a taint and automatically populate the operator given what is given

File: python-lib/dku_utils/taints.py (hunk 2 above)
Comment (vrutz) [reply python-lib/dku_utils/taints.py:53]: ya it turns out you're not allowed to have a static method and a class method with the same name in python.

**Comment by vrutz** [comment]
Also tested that the autoscaler works by forcing load on a lot of pods onto the cluster and I did see the autoscaler updating the capacity

**Review by amandineslx — APPROVED** [review]
OK for me, just update the changelog

### Tested

* Deploy DSS 12.6.0 via FM in elastic fleet
* Update plugin with archive built from PR (commit `b46ba4a6686e12bd8559d83f34689fb42ffc1c15`)
* Create cluster with: 1 node pool simple, 1 node pool with GPU, taints and label, 1 node pool with autoscaler, label and taints
  * ✅ node pool IDs are correctly set
  * ✅ autoscaler is deployed
  * ✅ nvidia daemon set is present for the node pools where the toleration is set and the one without taint but not on the one with other taint
  * ✅ nvidia daemon set version is correct
  * ✅ labels are populated correctly
  * ✅ taints are populated correctly
  * ✅ all 3 node pools are created in the cluster
  * :white_check_mark: autoscaler has the correct tolerations to be scheduled on the node pool where it should be
* Create cluster with: 1 node pool with autoscaler without taints ; add a new node pool with autoscaler and taints
  * :white_check_mark: tags representing the autoscaler tolerations are correctly added to the new
* ❌ release notes for next version in changelog

**Comment by amandineslx** [comment]
all good for me

**Pull request merged**